server/service/src/processors/transfer/requisition/mod.rs

```diff
@@ -0,0 +1,190 @@
+pub(crate) mod update_request_requisition_status;
+
+#[cfg(test)]
+pub(crate) mod test;
+
+use repository::{
+    ChangelogAction, ChangelogFilter, ChangelogRepository, ChangelogRow, ChangelogTableName,
+    EqualFilter, KeyValueStoreRepository, KeyValueType, RepositoryError, Requisition,
+    RequisitionFilter, RequisitionRepository, StorageConnection,
+};
+use thiserror::Error;
+
+use crate::{
+    processors::transfer::requisition::{
+        create_response_requisition::CreateResponseRequisitionProcessor,
+        link_request_requisition::LinkeRequestRequisitionProcessor,
+        update_request_requisition_status::UpdateRequestRequstionStatusProcessor,
+    },
+    service_provider::ServiceProvider,
+    sync::{ActiveStoresOnSite, GetActiveStoresOnSiteError},
+};
+
+const CHANGELOG_BATCH_SIZE: u32 = 20;
+
+#[derive(Clone, Debug)]
+pub(crate) struct RequisitionTransferProcessorRecord {
+    requisition: Requisition,
+    linked_requisition: Option<Requisition>,
+    other_party_store_id: String,
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum ProcessRequisitionTransfersError {
+    #[error("Problem getting upsert record {0}")]
+    GetUpsertRecordError(GetUpsertRecordError),
+    #[error("{0}")]
+    GetActiveStoresOnSiteError(GetActiveStoresOnSiteError),
+    #[error("{0:?}")]
+    DatabaseError(RepositoryError),
+    #[error("{0}")]
+    ProcessorError(ProcessorError),
+    #[error("Name id is missing from requisition changelog {0:?}")]
+    NameIdIsMissingFromChangelog(ChangelogRow),
+    #[error("Name is not an active store {0:?}")]
+    NameIsNotAnAciveStore(ChangelogRow),
+}
+
+// TODO transaction
+
+pub(crate) fn process_requisition_transfers(
+    service_provider: &ServiceProvider,
+) -> Result<(), ProcessRequisitionTransfersError> {
+    use ProcessRequisitionTransfersError as Error;
+    let processors: Vec<Box<dyn RequisitionTransferProcessor>> = vec![
+        Box::new(CreateResponseRequisitionProcessor),
+        Box::new(LinkeRequestRequisitionProcessor),
+        Box::new(UpdateRequestRequstionStatusProcessor),
+    ];
+
+    let ctx = service_provider.context().map_err(Error::DatabaseError)?;
+
+    let active_stores =
+        ActiveStoresOnSite::get(&ctx.connection).map_err(Error::GetActiveStoresOnSiteError)?;
+
+    let changelog_repo = ChangelogRepository::new(&ctx.connection);
+    let key_value_store_repo = KeyValueStoreRepository::new(&ctx.connection);
+    let filter = ChangelogFilter::new()
+        .table_name(ChangelogTableName::Requisition.equal_to())
+        .name_id(EqualFilter::equal_any(active_stores.name_ids()));
+
+    loop {
+        let cursor = key_value_store_repo
+            .get_i64(KeyValueType::RequisitionTransferProcessorCursor)
+            .map_err(Error::DatabaseError)?
+            .unwrap_or(0) as u64;
+
+        let logs = changelog_repo
+            .changelogs(cursor, CHANGELOG_BATCH_SIZE, Some(filter.clone()))
+            .map_err(Error::DatabaseError)?;
+
+        if logs.is_empty() {
```

Comment:
I wonder if we should break after a fixed 'MAX RECORDS' as well?
Just wondered if that might help if we had an infinite loop situation?
I'm just imagining some kind of loop forever situation that we can't easily see in the logs?

Reply:
Related, it might be good to count and log the number of records processed for debugging in future. In normal operations i'm expecting it to process a very small number, so if we see large numbers in the logs we'd know something wasn't quite right?

Reply:
Added this comment to https://github.com/openmsupply/open-msupply/issues/156#issuecomment-1221850743, under checkbox for extra protection, we can do this in another issue i think. I was also thinking that around here would be the best place to add extra protection/logging. (I have quite a lot more to add on this subject, but will wait until that issue)

Reply:
Outer loop is needed because RequestRequisition is not immediately `linked` to ResponseRequisition when ResponseRequisition is created when both request and response store are on the same site. In this case what happens is: ResponseRequisition is created, it's now in changelog, then outer loop captures that ResponseRequisition might need to be processed and runs the link processor.

Reply:
> If my understandering is correct then there is still a possible risk of failing to process creating transfers if omSupply crashed while several runs are queued - running all the processors on startup would address that though.

Hmm I've added re run on startup to [improvement section in this issue](https://github.com/openmsupply/open-msupply/issues/608)

Reply:
> Outer loop is needed because RequestRequisition is not immediately `linked` to ResponseRequisition when ResponseRequisition is created when both request and response store are on the same site. In this case what happens is: ResponseRequisition is created, it's now in changelog, then outer loop captures that ResponseRequisition might need to be processed and runs the link processor.

Oh right so the processor itself has updated records in the change log and needs to process those too.

Reply:
> As an aside, is CHANGELOG_BATCH_SIZE necessary

Is it bad to have it ?

Reply:
> > As an aside, is CHANGELOG_BATCH_SIZE necessary
> 
> Is it bad to have it ?

Nah it's probably all G, but in the insane 10,000 requisitions scenario you'll do 500 queries on the the database to process them 20 at a time rather than idk, 2 or 3 queries.

Reply:
In that scenario, we would also be doing 10,000 updates to the cursor row in the key value store too.

If we're optimising for large scale there's likely more important performance improvements we could make,and updating the batch size is easy if required in future.

Reply:
I agree, and I should clarify this is a nit (hence approving). I will highlight things that don't have apparent purpose in PRs - I've come from the 4D code base which has _large_ amounts of code that serves no concrete purpose that makes it much harder to read and maintain. Code that solves a problem that doesn't exist, then breaks, then code fixing the code that didn't need exist in the first place. It's much easier to nip it in the bud before 6 layers of complexity have made it worse.

Reply:
Just a side note, i am pretty sure in 4d `selection` and `entity selection` is not actually `loaded` into memory, just the cursors for those records, i think if you convert them to object you might have issues with memory in 4d. I guess what i am trying to say is I think there some purpose for the batch size, protection against loading too much into memory (although just theoretical benefit and not a conscious decision)